src/mbgl/mtl/context.cpp

```diff
@@ -0,0 +1,199 @@
+#include <mbgl/shaders/mtl/shader_program.hpp>
+#include <mbgl/util/traits.hpp>
+#include <mbgl/util/std.hpp>
+#include <mbgl/util/logging.hpp>
+
+#include <Foundation/Foundation.hpp>
+#include <Metal/Metal.hpp>
+
+#include <algorithm>
+#include <cstring>
+
+namespace mbgl {
+namespace mtl {
+
+Context::Context(RendererBackend& backend_)
+    : gfx::Context(16), // TODO
+      backend(backend_),
+      stats() {}
+
+Context::~Context() noexcept {
+    if (cleanupOnDestruction) {
+        performCleanup();
+        assert(stats.isZero());
+    }
+}
+
+std::unique_ptr<gfx::CommandEncoder> Context::createCommandEncoder() {
+    return std::make_unique<CommandEncoder>(*this);
+}
+
+BufferResource Context::createBuffer(const void* data, std::size_t size, gfx::BufferUsageType) const {
+    return {backend.getDevice(), data, size, MTL::ResourceStorageModeShared};
+}
+
+UniqueShaderProgram Context::createProgram(std::string name,
+                                           const std::string_view source,
+                                           const std::string_view vertexName,
+                                           const std::string_view fragmentName,
+                                           const ProgramParameters& programParameters,
+                                           const std::unordered_map<std::string, std::string>& additionalDefines) {
+    const auto pool = NS::TransferPtr(NS::AutoreleasePool::alloc()->init());
+
+    // No NSMutableDictionary?
+    const auto& programDefines = programParameters.getDefines();
+    const auto numDefines = programDefines.size() + additionalDefines.size();
+
+    std::vector<NS::Object*> rawDefines;
+    rawDefines.reserve(2 * numDefines);
+    const auto addDefine = [&rawDefines](const auto& pair) {
+        auto* nsKey = NS::String::string(pair.first.data(), NS::UTF8StringEncoding);
+        auto* nsVal = NS::String::string(pair.second.data(), NS::UTF8StringEncoding);
+        rawDefines.insert(std::next(rawDefines.begin(), rawDefines.size() / 2), nsKey);
+        rawDefines.insert(rawDefines.end(), nsVal);
+    };
+    std::for_each(programDefines.begin(), programDefines.end(), addDefine);
+    std::for_each(additionalDefines.begin(), additionalDefines.end(), addDefine);
+
+    const auto nsDefines = NS::Dictionary::dictionary(
+        &rawDefines[numDefines], rawDefines.data(), static_cast<NS::UInteger>(numDefines));
+    rawDefines.clear();
+
+    auto options = MTL::CompileOptions::alloc()->init();
+    options->setPreprocessorMacros(nsDefines);
+    options->setFastMathEnabled(true);
+    options->setLanguageVersion(MTL::LanguageVersion3_0);
+    options->setLibraryType(MTL::LibraryTypeExecutable);
+    options->setPreserveInvariance(true);
+    options->setOptimizationLevel(MTL::LibraryOptimizationLevelDefault);
+    options->setCompileSymbolVisibility(MTL::CompileSymbolVisibilityDefault);
+    options->setAllowReferencingUndefinedSymbols(false);
```

Comment:
These 3 functions are available starting with iOS 16, so we should remove them. Or better we should remove the entire `options` block, I don't see where it is used.

Reply:
Good catch on it not being used, that's just a bug.  We may want to make the optimization level available to improve initial load time, we just need to make it conditional on the target.